invokeai/backend/stable_diffusion/extensions_manager.py

```diff
@@ -0,0 +1,167 @@
+
+                else:
+                    raise Exception(f"Unsupported injection type: {inj_info.type}")
+
+    def call_callback(self, name: str, *args, **kwargs):
+        if name in self._callbacks:
+            self._callbacks[name](*args, **kwargs)
+
+    # TODO: is there any need in such high abstarction
+    # @contextmanager
+    # def patch_extensions(self):
+    #    exit_stack = ExitStack()
+    #    try:
+    #        for ext in self.extensions:
+    #            exit_stack.enter_context(ext.patch_extension(self))
+    #
+    #        yield None
+    #
+    #    finally:
+    #        exit_stack.close()
+
+    @contextmanager
+    def patch_attention_processor(self, unet: UNet2DConditionModel, attn_processor_cls: object):
+        unet_orig_processors = unet.attn_processors
+        exit_stack = ExitStack()
+        try:
+            # just to be sure that attentions have not same processor instance
+            attn_procs = {}
+            for name in unet.attn_processors.keys():
+                attn_procs[name] = attn_processor_cls()
+            unet.set_attn_processor(attn_procs)
+
+            for ext in self.extensions:
+                exit_stack.enter_context(ext.patch_attention_processor(attn_processor_cls))
+
+            yield None
+
+        finally:
+            unet.set_attn_processor(unet_orig_processors)
+            exit_stack.close()
+
+    @contextmanager
+    def patch_unet(self, state_dict: Dict[str, torch.Tensor], unet: UNet2DConditionModel):
+        exit_stack = ExitStack()
+        try:
+            changed_keys = set()
+            changed_unknown_keys = {}
+
+            for ext in self.extensions:
+                patch_result = exit_stack.enter_context(ext.patch_unet(state_dict, unet))
+                if patch_result is None:
+                    continue
+                new_keys, new_unk_keys = patch_result
+                changed_keys.update(new_keys)
+                # skip already seen keys, as new weight might be changed
+                for k, v in new_unk_keys.items():
+                    if k in changed_unknown_keys:
+                        continue
+                    changed_unknown_keys[k] = v
+
+            yield None
+
+        finally:
+            exit_stack.close()
+            assert hasattr(unet, "get_submodule")  # mypy not picking up fact that torch.nn.Module has get_submodule()
+            with torch.no_grad():
+                for module_key in changed_keys:
+                    weight = state_dict[module_key]
+                    unet.get_submodule(module_key).weight.copy_(
+                        weight, non_blocking=TorchDevice.get_non_blocking(weight.device)
+                    )
+                for module_key, weight in changed_unknown_keys.items():
+                    unet.get_submodule(module_key).weight.copy_(
+                        weight, non_blocking=TorchDevice.get_non_blocking(weight.device)
+                    )
```

Comment:
Same comment as above. This patch_unet() implementation is too opinionated. It should just call down to the extensions and let them handle things like weight restoration.

E.g. what if we wanted a series of patches that: 1) modify weights, 2) change shape, 3) modify weights more. This implementation makes too many assumptions to enable that.

Reply:
A standard pattern for patching is to have patches act indepenedently and then ensure that they are un-patched in the reverse order:

1. Apply patch 1
2. Apply patch 2
3. Apply patch 3
4. Do something
5. Unpatch 3
6. Unpatch 2
7. Unpatch 1

This is great. As long as each patch can undo itself, this is guaranteed to work without any conflicts between patches.

The current implementation breaks this pattern, so introduces opportunities for conflicts between patches.

I'm not sure that I fully understand how the current implementation is intended to work. Can you clarify the intended difference between `new_keys` and `new_unk_keys`?
My concern is that if you mix patches that change the model's shape with ones that just change weights, you will end up with situations where it's hard to track which weights were the originals. But, maybe all we need here is better documentation around exactly what types of changes patches need to *report*.

Also, I've just realized that the `model_state_dict` param to apply_lora(...) really throws a wrench into all of this...

Reply:
From completely abstract point of view - it's how we should do.
But as we working with tensors and we know that we want 2 moments - get patched model and later revert to original, we can gain some more suitable requirements to code.
1) There no need in intermediate weights, as we interested only in original and fully patched models.
Even if extension will require them for some reasons - it can simply save them and later use.
2) Except for data(weight) we also have shapes and module modifications themselves(like custom forward in `seamless`) this field a bit trickier and this not generate any memory impact, so this part I left to be done by extensions.

So there shouldn't be any problems. Else if each patcher will store all intermediate weights we will quickly notice memory impact, while there really no reason to store this data.

About keys - each patcher returns weights which it modifies. 
If this key presented in `model_state_dict` this weight's key will be added to `keys`.
Else - patcher clones this weight and sends key+weight in `unk_keys`.

In `patch_unet` with `keys` we can simply add them as there no weight provided and wrong restore is impossible.
With `unk_keys` we store only first appearance of each key, as second appearance means that patcher returns already patched version.

About tracking - how is this possible? About report - any weight modification should be reported, and shape here not affect anything as for keys from `model_state_dict` we simply restore weights by key, no extra info from patcher, for unk_keys(missing in `model_state_dict`) we store only firstly reported one and it always should be cloned before applying any patches to it, so by this we always get here original weights.